File: c/bfac.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LEN     9

int valid_factoradic(const char *fac)
{
    int d;
    const char *p;

    d = 0;
    p = fac + strlen(fac) - 1;

    for (d = 0, p = fac + strlen(fac) - 1; p >= fac; d++, p--) {
        if ((*p - '0') > d) {
            return 0;
        }
    }

    return 1;
}

int fac2ten(const char *fac)
{
    int n, f, len;
    const char *p;

    len = strlen(fac);
    if (len > MAX_LEN) {
        return -1;
    }

    n = 0;
    f = 1;
    p = fac + len - 2;

    while (p >= fac) {
        n += (*p - '0') * f;
        f *= (f + 1);
        --p;
    }

    return n;
}

static int ten2fac_r(int n, int f, char *fac)
{
    int d;
    char s[2];

    if (n < f) {
        return n;
    }

    n = ten2fac_r(n, f * (f + 1), fac);

    d = n / f;
    n %= f;

    sprintf(s, "%d", d);
    strcat(fac, s);

    return n;
}

char *ten2fac(int n)
{
    static char fac[MAX_LEN+1];

    fac[0] = '\0';
    ten2fac_r(n, 1, fac);
    strcat(fac, "0");

    return fac;
}

File: python/bfac.py
# requires python2.6 due to use of factorial function

from math import factorial

def valid_factoradic(number):
	return True

def ten2fac(bten):
	"""
	Convert a number from base ten to base factorial.

	The result is an actual integer, no subscript attached.
	
	>>> [ten2fac(n) for n in range(24)]
	[0, 10, 100, 110, 200, 210, 1000, 1010, 1100, 1110, 1200, 1210, 2000, 2010, 2100, 2110, 2200, 2210, 3000, 3010, 3100, 3110, 3200, 3210]
	"""

	num = ''

	fac = 0
	while bten >= factorial(fac):
		fac = fac + 1

	if fac > 0:
		fac = fac - 1

	radix = r = fac + 1

	t = bten
	while radix:
		f = factorial(fac)
		digit = t / f
		t = t % f
		fac = fac - 1
		radix = radix - 1

		num = num + str(digit)

	if len(num) < radix:
		num = num + '0'

	return int(num)

def fac2ten(bfac):
	"""
	Convert a number from base factorial to base ten.

	>>> [fac2ten(n) for n in 0, 10, 100, 110, 200, 210, 1000, 1010, 1100, 1110, 1200, 1210, 2000, 2010, 2100, 2110, 2200, 2210, 3000, 3010, 3100, 3110, 3200, 3210]
	[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
	"""

	if not valid_factoradic(bfac):
		error = 'The number ' + str(bfac) + ' is not a valid factorial number'
		raise TypeError(error)

	num  = 0

	f = str(bfac)

	radix = len(f)

	pos = 0
	while radix > 1:
		radix = radix - 1
		v = int(f[pos])
		n = v * factorial(radix)
		num = num + n
		pos = pos + 1

	return num
